codegen/masm/tests/test_basic_arithmetic.rs

```diff
@@ -0,0 +1,227 @@
+use air_codegen_masm::{code_gen, constants};
+use assembly::Assembler;
+use ir::AirIR;
+use miden_diagnostics::{
+    term::termcolor::ColorChoice, CodeMap, DefaultEmitter, DiagnosticsHandler,
+};
+use parser::{ast::Source, Parser};
+use processor::{AdviceInputs, Kernel, MemAdviceProvider, Process, StackInputs};
+use std::sync::Arc;
+
+pub fn parse(source: &str) -> Source {
+    let codemap = Arc::new(CodeMap::new());
+    let emitter = Arc::new(DefaultEmitter::new(ColorChoice::Auto));
+    let diagnostics = DiagnosticsHandler::new(Default::default(), codemap.clone(), emitter);
+    let parser = Parser::new((), codemap);
+
+    parser
+        .parse_string::<Source, _, _>(&diagnostics, source)
+        .expect("parsing failed")
+}
+
+static ARITH_AIR: &str = "
+def Add
+
+trace_columns:
+    main: [a, b]
+
+public_inputs:
+    stack_inputs: [16]
+
+boundary_constraints:
+    enf a.first = 0
+
+integrity_constraints:
+    enf a + a = 0
+    enf a - a = 0
+    enf a * a = 0
+    enf a^3 = 0
+
+    enf b + a = 0
+    enf b - a = 0
+    enf b * a = 0
+";
+
+#[test]
+fn test_simple_arithmetic() {
+    let ast = parse(ARITH_AIR);
+    let ir = AirIR::new(ast).expect("build AirIR failed");
+    let code = code_gen(&ir).expect("codegen failed");
+
+    let (a, b) = (3u64, 7u64);
+    let code = format!(
+        "
+        begin
+            push.{}          # frame address
+            push.0.0.{b}.{a} # data
+            dup.4 mem_storew # save data to memory
+            dropw drop       # clean stack
+
+            # [generated code]
+            {}
```

Comment:
The generated code isn't pretty:

```
            padw mem_loadw.4294965000 movdn.3 drop drop drop padw mem_loadw.4294965000 movdn.3 drop drop drop add push.0 sub padw mem_loadw.4294965000 movdn.3 drop drop drop padw mem_loadw.4294965000 movdn.3 drop drop drop sub push.0 sub padw mem_loadw.4294965000 movdn.3 drop drop drop padw mem_loadw.4294965000 movdn.3 drop drop drop mul push.0 sub padw mem_loadw.4294965000 movdn.3 drop drop drop exp.3 push.0 sub padw mem_loadw.4294965000 drop movdn.2 drop drop padw mem_loadw.4294965000 movdn.3 drop drop drop add push.0 sub padw mem_loadw.4294965000 drop movdn.2 drop drop padw mem_loadw.4294965000 movdn.3 drop drop drop sub push.0 sub padw mem_loadw.4294965000 drop movdn.2 drop drop padw mem_loadw.4294965000 movdn.3 drop drop drop mul push.0 sub
```

AFAICT there is no easy way to go from constraint root to the air expression, which would be nice to emit as comments. There is also no easy way to go backwards from column index to variable name, so the results ends up being the blob above.

Each value is loaded individually since it was agreed at this stage we don't want to have optimizations (and that would need something akin to the middle IR, which we decided not to do).

Reply:
A few thoughts on this:

1. We should probably emit the code as several procedures. At the very least, I'd have 3 procedures: one for each type of constraint. For example, `eval_boundary_first`, `eval_boundary_last`, and `eval_transition`. These 3 procedures could be executed sequentially by a wrapper procedure and then the results could be combined by a boilerplate procedure. Having this structure would make the results a bit easier to debug.
2. I'd separate the emitted code for each constraint by a newline. Also, even if we can't link back to the original expression, I'd probably still emit comments to separate one constraint from another (i.e., `# constraint 1`, `# constraint 2` etc.).
3. There is no need to optimize at this point, but I think once we have the general structure working, we'd probably need to spend some time on optimizations to get the cycle count of Miden VM constraints to (ideally) under 30K cycles.

Reply:
1. Agreed. This only implements the translation of AirIR to Masm code, which should be the same for transition and boundary constraints. Wrapping this in a procedure has been agreed upon and should be easy to add on top of this.
2. Okay, I will change the vector to instead of having instructions only, it will contain MASM fragments, to include the comments you suggested.
3. Left some comments on optimization on a previous comment.

Reply:
Updated the codegen, now it adds a comment:

```
 padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop ext2add push.0.0 ext2sub # constraint 0
 padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop ext2sub push.0.0 ext2sub # constraint 1
 padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop ext2mul push.0.0 ext2sub # constraint 2
 padw mem_loadw.4294965000 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop ext2add push.0.0 ext2sub # constraint 3
 padw mem_loadw.4294965000 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop ext2sub push.0.0 ext2sub # constraint 4
 padw mem_loadw.4294965000 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop ext2mul push.0.0 ext2sub # constraint 5
```

Reply:
I changed the codegen to introduce nested control flow and add procedures, now it looks like this:

```
proc.evaluate_transitions
    exec.cache_periodic_polys
    padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop padw mem_loadw.500000000 drop drop ext2mul push.0 push.0 ext2sub # constraint 0 for main
end # END PROC evaluate_transitions
```